domain/operation/service/start.go

```diff
@@ -143,6 +175,37 @@ func (s *Service) AddExecOperation(
 		}
 	}
 
+	// Make sure that we have the same number of result slots as the
+	// pre-computed results. This is a sanity check.
+	if len(runResult.Machines) > len(existingMachines) {
+		s.logger.Errorf(ctx, "more state layer results than result slots: %d > %d", len(runResult.Units), len(existingMachines))
+		// This should never happen, but if it does, we'll just truncate the
+		// state layer results to match the result slots.
+		runResult.Machines = runResult.Machines[:len(existingMachines)]
+	}
```

Comment:
If this should never happen, then return an error. Don't make it worse by making things non-deterministic.